stdenv: fix crossSystem localSystem comparison

Artturin opened the pull request:
this appears to fix the following issue for some reason

```
nix-repl> (import ./. { localSystem.system = "x86_64-linux"; crossSystem.system = "x86_64-linux"; }).bash
«derivation /nix/store/5lz9p8xhf89kb1c1kk6jxrzskaiygnlh-bash-5.2-p15.drv»

nix-repl> (import ./. { localSystem.system = "x86_64-linux"; crossSystem = "x86_64-linux"; }).bash
«derivation /nix/store/1whiq03rsfrmch2ds3jhyqjaczfz97lb-bash-5.2-p15-x86_64-unknown-linux-gnu.drv»
```

fixed
```
nix-repl> (import ./. { localSystem.system = "x86_64-linux"; crossSystem = "x86_64-linux"; }).bash
«derivation /nix/store/5lz9p8xhf89kb1c1kk6jxrzskaiygnlh-bash-5.2-p15.drv»
```

issue reported by @spikespaz on matrix in the cross compilation channel 

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Artturin committed:
stdenv: fix crossSystem localSystem comparison

this appears to fix the following issue for some reason

```
nix-repl> (import ./. { localSystem.system = "x86_64-linux"; crossSystem.system = "x86_64-linux"; }).bash
«derivation /nix/store/5lz9p8xhf89kb1c1kk6jxrzskaiygnlh-bash-5.2-p15.drv»

nix-repl> (import ./. { localSystem.system = "x86_64-linux"; crossSystem = "x86_64-linux"; }).bash
«derivation /nix/store/1whiq03rsfrmch2ds3jhyqjaczfz97lb-bash-5.2-p15-x86_64-unknown-linux-gnu.drv»
```

fixed
```
nix-repl> (import ./. { localSystem.system = "x86_64-linux"; crossSystem = "x86_64-linux"; }).bash
«derivation /nix/store/5lz9p8xhf89kb1c1kk6jxrzskaiygnlh-bash-5.2-p15.drv»
```

roberth commented:
Ohh I did not expect to see this problem in practice, having noticed it only while handling platforms manually and not through the nixpkgs entrypoint, for testing.
I'd discussed a plan to mitigate the problem with @Ericson2314. It's still more of a workaround than a solution, I have to admit.
1. add a platform equality function to `lib.systems`
2. use it in all places where nixpkgs consumers system/platform parameters, so they can be turned into equal platforms, along these lines

```nix
# *Platform omitted for brevity
args@{ host, build }:
let inherit (
  # actually this part could be turned into a function
  if systems.equals args.host args.build
  then { inherit (args) host; build = host; }
  else { inherit (args) host build; }
  ) host build;
in
  # use host and build as usual
```

That way we can keep using regular equality on these types.
What's ugly is that its a bad habit in case you ever have to compare systems between distinct nixpkgs invocations. There's no workaround for that, except to use `lib.systems.equals`.

I think the real solution is to either
a. make equality break, so that nobody can use it by accident, or
b. remove all functions and put them in `lib.systems` instead.

But until we decide that, we may as well fix it for 95% of cases by explicitly making the values identical where it matters most.

spikespaz commented:
The issue happens simply when just importing `nixpkgs` with `localSystem.system = "x86_64-linux"` or `localSystem = "x86_64-linux"`. Whether or not I specify `crossSystem = localSystem` makes no difference; I have tried it both ways, traced it through `default.nix` and `top-level/impure.nix` and have determined that this would be the default regardless. I don't know why nobody else can confirm this, as the broken example given in the issue description is not as simple as my use.

Perhaps we could fix 100% of the problem by fixing function comparison in Nix?

```nix
  intrepid = inputs.nixpkgs.lib.nixosSystem {
    modules = with tree.hosts.intrepid; [
      bootloader
      filesystems
      plymouth
      configuration
      packages
      powerplan
      touchpad
      greeter
      # gamemode
      gaming
      pia-openvpn
    ];
    pkgs = import inputs.nixpkgs-unstable {
      overlays = [
        # flake packages
        self.overlays.default
        # override packages with an unfree license
        self.overlays.allowUnfree
        # other packages
        inputs.slight.overlays.default
        inputs.ragenix.overlays.default
      ];
      localSystem = "x86_64-linux";
      crossSystem = "x86_64-linux";
      config.allowUnfree = true;
    };
    specialArgs = {
      enableUnstableZfs = false;
    };
  };
  ```

roberth commented:
> Perhaps we could fix 100% of the problem by fixing function comparison in Nix?

It's fundamentally unfixable because of undecidability. We _should_ make it throw, but that's a big breaking change.
I suppose we could make equality overridable. That would still take a while to be available in Nixpkgs, fwiw.

> ```
> error: Neither nixpkgs.hostPlatform nor the legacy option nixpkgs.system has been set.
> ```

That shouldn't happen. You've specified `pkgs` manually, so the options in `misc/nixpkgs.nix` should be ignored. However, it is set through `_module.args.pkgs` and not through `nixpkgs.pkgs` so `misc/nixpkgs.nix` doesn't know that it should disable its assertions.

I've added a safe alternative read-only nixpkgs module as part of this.
 - https://github.com/NixOS/nixpkgs/pull/230523

Maybe that should be loaded by `nixosSystem`/`eval-config.nix` when `pkgs` is set, or maybe its behavior should be integrated into the standard module. That's a quite a challenge, not to break existing configs while making subtle changes there, but if we're willing to break the `pkgs` argument behavior anyway...

For now you could fix your example by setting the `nixpkgs.pkgs` option instead of the `nixosSystem` `pkgs` argument.

roberth commented on pkgs/stdenv/default.nix:
```diff
@@ -34,9 +34,11 @@ let
 
   stagesCustom = import ./custom args;
 
+  removeFunctions = a: lib.filterAttrs (_: v: !builtins.isFunction v) a;
```
Added your solution with the test I previously wrote (almost unchanged)
- #237512

spikespaz commented:
> For now you could fix your example by setting the nixpkgs.pkgs option instead of the nixosSystem pkgs argument.

This seems to me missing the binary cache, I am rebuilding too many things. Tried both `localSystem.system =` and `localSystem =`.

Artturin committed:
treewide: use lib.systems.equals to compare systems

spikespaz commented:
I have cherry-picked the commits from both pull requests, but get infinite recursion. Now when I try to add `--show-trace` to the command, I see
```
error: cached failure of attribute 'nixosConfigurations.intrepid.config'
```
This is happening a lot now and it is very annoying. Sometimes I want to see an error message multiple times. How do I prevent this "cache" error from happening? What does it mean? Why am I seeing it more?

spikespaz commented:
Is it possible that the following lines in `impure.nix` are responsible for the original problem?
My guess is that the values `localSystem` and `crossSystem` are not the same if both are provided, as some other value is used via the confusing default logic.

This is the top of the attrset bound to `args`:
```nix
{ # We put legacy `system` into `localSystem`, if `localSystem` was not passed.
  # If neither is passed, assume we are building packages on the current
  # (build, in GNU Autotools parlance) platform.
  localSystem ? { system = args.system or builtins.currentSystem; }

# These are needed only because nix's `--arg` command-line logic doesn't work
# with unnamed parameters allowed by ...
, system ? localSystem.system
, crossSystem ? localSystem
```

Perhaps the issue that you "did not expect to see ... in practice" is caused by this confusing evaluation flow.

My tests show otherwise, but I feel as if I am still missing something.

```nix
nix-repl> test = args @ { localSystem ? { system = args.system; }, system ? localSystem.system, crossSystem ? localSystem }: { inherit localSystem crossSystem; }

nix-repl> :p test { system = "x86_64-linux"; }
{ crossSystem = { system = "x86_64-linux"; }; localSystem = «repeated»; }

nix-repl> :p test { localSystem = "x86_64-linux"; }
{ crossSystem = "x86_64-linux"; localSystem = "x86_64-linux"; }

nix-repl> :p test { localSystem.system = "x86_64-linux"; }
{ crossSystem = { system = "x86_64-linux"; }; localSystem = «repeated»; }

nix-repl> :p test { localSystem = "x86_64-linux"; crossSystem = "x86_64-linux"; }
{ crossSystem = "x86_64-linux"; localSystem = "x86_64-linux"; }

nix-repl> :p test { localSystem.system = "x86_64-linux"; crossSystem.system = "x86_64-linux"; }
{ crossSystem = { system = "x86_64-linux"; }; localSystem = { system = "x86_64-linux"; }; }
```

roberth commented:
> error: cached failure of attribute 'nixosConfigurations.intrepid.config'

Which nix version, and which command is that?
That error should only ever be visible in commands that deal with large quantities of attributes, so this particular error seems like a Nix bug.

roberth committed:
pkgs/top-level: Improve crossSystem equality early on

This is a workaround that lets us use `==` on systems despite the
functions that would make `==` always return `false` if a
workaround like this is not applied.

The real solution is to either use `lib.systems.equals` for all
such comparisons, or to remove all functions from the elaborated
systems.

spikespaz commented:
> Which nix version, and which command is that?

Both `nix build` and `nixos-rebuild` produce the same message if invoked twice, the first time resulting in an error code and trace.

Nix version is `2.15.1`.

Artturin commented on pkgs/top-level/default.nix:
```diff
@@ -60,11 +60,24 @@ in let
-    if crossSystem0 == null || crossSystem0 == args.localSystem
-    then localSystem
-    else lib.systems.elaborate crossSystem0;
+    if crossSystem0 == null || lib.systems.equals crossSystem0 args.localSystem
```
https://gist.github.com/GrahamcOfBorg/1e800ef80702f76206b24b79c9f0dc1d

`error: value is a string while a set was expected`

crossSystem0 and args.localSystem haven't been elaborated so they're just strings

amjoseph-nixpkgs replied:
> crossSystem0 and args.localSystem haven't been elaborated so they're possibly just strings

IMHO we should elaborate unconditionally here.

amjoseph-nixpkgs commented:
> I think the real solution is to either
> a. make equality break, so that nobody can use it by accident, or
> b. remove all functions and put them in `lib.systems` instead.

I think these are the only two reasonable solutions.

Does anybody object strongly to (b)?

Artturin commented:
https://github.com/NixOS/nixpkgs/pull/254763 was merged
